lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/IncrementalSummarizer.java

```diff
@@ -0,0 +1,311 @@
+
+        // Prune unsupported itemsets in the current pane
+        calcCumSum();
+        int currPanes = inlierPaneCounts.size();
+        for (Set<Integer> itemset : outlierItemsetWindowCount.keySet()) {
+            double supportSinceTracked = outlierCountCumSum.get(currPanes) -
+                    outlierCountCumSum.get(trackingMap.get(itemset));
+            if (outlierItemsetWindowCount.getOrDefault(itemset, 0.0) < minOutlierSupport * supportSinceTracked) {
+                unSupported.add(itemset);
+                trackingMap.remove(itemset);
+            }
+        }
+        outlierItemsetPaneCount.keySet().removeAll(unSupported);
+        inlierItemsetPaneCount.keySet().removeAll(unSupported);
+        outlierItemsetWindowCount.keySet().removeAll(unSupported);
+        inlierItemsetWindowCount.keySet().removeAll(unSupported);
+    }
+
+    /* Helper function that updates internals to keep track of a promoted itemset. */
+    private void trackItemset(Set<Integer> itemset, double count) {
+        trackingMap.put(itemset, inlierPaneCounts.size() - 1);
+        outlierItemsetWindowCount.put(itemset, count);
+        outlierItemsetPaneCount.put(itemset, count);
+    }
+
+    /* This function picks up itemsets that have enough outlier support in the current pane
+     * (a.k.a. itemsets that can be promoted) and starts tracking their occurances in the window. */
+    private void addNewFrequent() {
+        double minSupport = Math.ceil(minOutlierSupport * outlierItemsets.size());
+        HashMap<Integer, Double> inlierPaneSingletonCount = new ExactCount().count(inlierItemsets).getCounts();
+        // Get new frequent itemsets in outliers
+        FPGrowth fpGrowth = new FPGrowth();
+        List<ItemsetWithCount> frequent = fpGrowth.getItemsetsWithSupportCount(outlierItemsets,
+                minSupport);
+        List<ItemsetWithCount> newFrequent = new ArrayList<>();
+        for (ItemsetWithCount iwc : frequent) {
+            Set<Integer> itemset = iwc.getItems();
+            if (!outlierItemsetWindowCount.containsKey(itemset)) {
+                trackItemset(itemset, iwc.getCount());
+                newFrequent.add(iwc);
+            }
+        }
+        // Get support in the inlier transactions
+        List<ItemsetWithCount> inlierCount = fpGrowth.getCounts(
+                inlierItemsets, inlierPaneSingletonCount, inlierPaneSingletonCount.keySet(), newFrequent);
+        for (ItemsetWithCount iwc : inlierCount) {
+            inlierItemsetWindowCount.put(iwc.getItems(), iwc.getCount());
+            inlierItemsetPaneCount.put(iwc.getItems(), iwc.getCount());
+        }
+    }
+
+    @Override
+    public void process(DataFrame df) {
+        // 1. Retire old pane counts if necessary
+        if (inlierPaneCounts.size() == numPanes) { expireLastPane(); }
+        // 2. Add support counts for the new pane
+        encodeAttributes(df);
+        addNewPane();
+        if (outlierItemsets.size() * minOutlierSupport >= 1) {
```

Comment:
Clean up the logic here: I think pruning should always happen, for addNewFrequent we may need to be more careful.